plugins/realtek-mst/fu-realtek-mst-device.c

```diff
@@ -0,0 +1,908 @@
+		return NULL;
+	if (!flash_iface_read (self, bank_address, image_bytes, FLASH_USER_SIZE, error))
+		return NULL;
+	return fu_firmware_new_from_bytes(g_bytes_new_take (g_steal_pointer (&image_bytes), FLASH_USER_SIZE));
+}
+
+static GBytes*
+fu_realtek_mst_device_dump_firmware (FuDevice *device, GError **error)
+{
+	FuRealtekMstDevice *self = FU_REALTEK_MST_DEVICE (device);
+	g_autofree void *flash_contents = g_malloc(FLASH_SIZE);
+
+	if (!mst_ensure_device_address (self, I2C_ADDR_ISP, error))
+		return NULL;
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_READ);
+	if (!flash_iface_read (self, 0, flash_contents, FLASH_SIZE, error))
+		return NULL;
+	fu_device_set_status (device, FWUPD_STATUS_IDLE);
+
+	return g_bytes_new_take (g_steal_pointer (&flash_contents), FLASH_SIZE);
+}
+
+static gboolean
+fu_realtek_mst_device_attach (FuDevice *device, GError **error)
+{
+	FuRealtekMstDevice *self = FU_REALTEK_MST_DEVICE (device);
+	guint8 value;
+
+	if (!mst_ensure_device_address (self, I2C_ADDR_ISP, error))
+		return FALSE;
+
+	/* re-enable hardware write protect via GPIO */
+	if (!mst_set_gpio88 (self, 0, error))
+		return FALSE;
+
+	if (!mst_read_register (self, REG_MCU_MODE, &value, error))
+		return FALSE;
+	if ((value & MCU_MODE_ISP) != 0) {
+		g_debug ("resetting device to exit ISP mode");
+		fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+
+		/* Set register EE bit 2 to request reset. This write can fail
+		 * spuriously, so we ignore the write result and verify the device is
+		 * no longer in programming mode after giving it time to reset. */
+		if (!mst_read_register (self, 0xEE, &value, error))
+			return FALSE;
+		mst_write_register (self, 0xEE, value | 2, NULL);
+
+		/* allow device some time to reset */
+		g_usleep(G_TIME_SPAN_SECOND);
+
+		/* verify device has exited programming mode and actually reset */
+		if (!mst_read_register (self, REG_MCU_MODE, &value, error))
+			return FALSE;
+		if ((value & MCU_MODE_ISP) == MCU_MODE_ISP) {
+			g_set_error_literal (error, FWUPD_ERROR,
+					     FWUPD_ERROR_NEEDS_USER_ACTION,
+					     "device failed to reset when requested");
+			fu_device_add_flag (device, FWUPD_DEVICE_FLAG_NEEDS_SHUTDOWN);
+			return FALSE;
+		}
+	} else {
+		g_debug ("device is already in normal mode");
+	}
+
+	fu_device_remove_flag (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
+	fu_device_set_status (device, FWUPD_STATUS_IDLE);
+	return TRUE;
+}
+
+static void
+fu_realtek_mst_device_init (FuRealtekMstDevice *self)
+{
+	FuDevice *device = FU_DEVICE (self);
+	self->dp_aux_dev_name = NULL;
+	self->bus_device = NULL;
+	self->active_bank = FLASH_BANK_INVALID;
+
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_INTERNAL);
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_DUAL_IMAGE);
+	fu_device_set_version_format (device, FWUPD_VERSION_FORMAT_PAIR);
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_CAN_VERIFY_IMAGE);
+
+	fu_device_add_protocol (device, "com.realtek.rtd2142");
+	fu_device_set_vendor (device, "Realtek");
```

Comment:
I think you probably also need a call for `fu_device_add_vendor_id`

Reply:
Moved from quirk file.